### ISSUE TITLE
[JENKINS-51198] Jenkins' CauseOfBlockage through the API

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#22732]
<p>Hi,</p>

<p> </p>

<p>It may sound like a duplicate of <del><a href="https://issues.jenkins.io/browse/JENKINS-41579" title="API for Pipeline level for causeOfBlockage" class="issue-link" data-issue-key="JENKINS-41579"><del>JENKINS-41579</del></a></del> but it's not.</p>

<p>It's currently not possible to have the <tt>CauseOfBlockage</tt> through the json API.</p>

<p>There are currently three half-baked ways to do this, but none are usable for me.</p>

<p>To reproduce, create a pipeline with the following groovy script:</p>

<pre>
node(<span class="code-quote">'build3'</span>) { <span class="code-comment">// change <span class="code-keyword">for</span> a name that you possess and one that you don't
</span>  stage(<span class="code-quote">'test'</span>) {
    sleep 5
  }
}
</pre>
<h1><a name="Firstway"></a>First way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/queue/api/json</tt> , it shows (among other) the following piece of data:</p>

<p> </p>

<pre>
{
      <span class="code-quote">"_class"</span>: <span class="code-quote">"hudson.model.Queue$BuildableItem"</span>,
      <span class="code-quote">"actions"</span>: [
{}
      ],
      <span class="code-quote">"blocked"</span>: <span class="code-keyword">false</span>,
      <span class="code-quote">"buildable"</span>: <span class="code-keyword">true</span>,
      <span class="code-quote">"id"</span>: 264,
      <span class="code-quote">"inQueueSince"</span>: 1525768721244,
      <span class="code-quote">"params"</span>: "",
      <span class="code-quote">"stuck"</span>: <span class="code-keyword">true</span>,
      <span class="code-quote">"task"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution$PlaceholderTask"</span>
      },
      <span class="code-quote">"url"</span>: <span class="code-quote">"queue/item/264/"</span>,
      <span class="code-quote">"why"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
      <span class="code-quote">"buildableStartMilliseconds"</span>: 1525768721245,
      <span class="code-quote">"pending"</span>: <span class="code-keyword">false</span>
},
</pre>
<p>The other <tt>why</tt> messages could be, for example:<br/>
 "why": "build2 is offline",<br/>
  </p>

<p>I believe all possible messages are located there: <a href="https://github.com/jenkinsci/jenkins/tree/1feb23e55ca807a664cf89986241155f24038aa5/core/src/main/resources/hudson/model/queue/CauseOfBlockage" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/tree/1feb23e55ca807a664cf89986241155f24038aa5/core/src/main/resources/hudson/model/queue/CauseOfBlockage</a></p>
<h3><a name="Theproblem"></a>The problem</h3>

<p>I cannot use this information for 2 reasons:</p>
<ol>
	<li>it is language dependent</li>
	<li><tt>BECAUSE_LABEL_IS_BUSY</tt> and <tt>BECAUSE_NODE_IS_BUSY</tt> result in the same string</li>
	<li>the message is for humans, not made for automation</li>
</ol>

<h1><a name="Secondway"></a>Second way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/{pipeline_name}/queue/</tt> , it should show the piece of data given in <a href="https://issues.jenkins-ci.org/browse/JENKINS-41579" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-41579</a> .</p>

<p>When I do it with my current queue (4 items in the queue), I see this:</p>

<pre>
[]
</pre>
<p>Not really useful, eh?</p>

<p>And with another pipeline, I see (among other lines) this:</p>

<pre>
<span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"Build #160 is already in progress (ETA:N/A)"</span>,
</pre>
<h3><a name="Theproblem"></a>The problem</h3>

<p>I think this one is self explanatory. However, there are other issues to it:</p>
<ol>
	<li>All queued items don't appear</li>
	<li>When it does appear, the message is for humans, not made for automation.</li>
	<li>It is pipeline specific. I couldn't find a way to have this info for all pipelines (or only the latest run within each pipeline)</li>
	<li>It relies on Blue Ocean to be installed</li>
</ol>

<p> </p>
<h1><a name="Thirdway"></a>Third way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/</tt>{<tt>pipeline_name</tt>}<tt>/runs/</tt> , it shows (among other) the following piece of data:</p>

<pre>
  {
    <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.PipelineRunImpl"</span>,
    <span class="code-quote">"_links"</span>: {
      <span class="code-quote">"parent"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/"</span>
      },
      <span class="code-quote">"tests"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/tests/"</span>
      },
      <span class="code-quote">"nodes"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/nodes/"</span>
      },
      <span class="code-quote">"log"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/log/"</span>
      },
      <span class="code-quote">"self"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/"</span>
      },
      <span class="code-quote">"actions"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/actions/"</span>
      },
      <span class="code-quote">"steps"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/steps/"</span>
      },
      <span class="code-quote">"artifacts"</span>: {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
<span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/artifacts/"</span>
      }
    },
    <span class="code-quote">"actions"</span>: [],
    <span class="code-quote">"artifactsZipFile"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
    <span class="code-quote">"causes"</span>: [
      {
<span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.service.embedded.<span class="code-keyword">rest</span>.AbstractRunImpl$BlueCauseImpl"</span>,
<span class="code-quote">"shortDescription"</span>: <span class="code-quote">"Started by user XXX"</span>,
<span class="code-quote">"userId"</span>: <span class="code-quote">"xxx"</span>,
<span class="code-quote">"userName"</span>: <span class="code-quote">"XXX"</span>
      }
    ],
    <span class="code-quote">"changeSet"</span>: [],
    <span class="code-quote">"description"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"durationInMillis"</span>: 0,
    <span class="code-quote">"enQueueTime"</span>: <span class="code-quote">"2018-05-08T08:38:37.482+0000"</span>,
    <span class="code-quote">"endTime"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"estimatedDurationInMillis"</span>: 102598,
    <span class="code-quote">"id"</span>: <span class="code-quote">"87"</span>,
    <span class="code-quote">"name"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"organization"</span>: <span class="code-quote">"jenkins"</span>,
    <span class="code-quote">"pipeline"</span>: <span class="code-quote">"{pipeline_name}"</span>,
    <span class="code-quote">"replayable"</span>: <span class="code-keyword">true</span>,
    <span class="code-quote">"result"</span>: <span class="code-quote">"UNKNOWN"</span>,
    <span class="code-quote">"runSummary"</span>: <span class="code-quote">"?"</span>,
    <span class="code-quote">"startTime"</span>: <span class="code-quote">"2018-05-08T08:38:37.486+0000"</span>,
    <span class="code-quote">"state"</span>: <span class="code-quote">"QUEUED"</span>,
    <span class="code-quote">"type"</span>: <span class="code-quote">"WorkflowRun"</span>,
    <span class="code-quote">"branch"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"commitId"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"commitUrl"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"pullRequest"</span>: <span class="code-keyword">null</span>
  },
</pre>
<p>We can see the following line:</p>

<pre>
<span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
</pre>
<p>So, here we have the information again.</p>
<h3><a name="Theproblem"></a>The problem</h3>

<p>It comes with several problems:</p>
<ol>
	<li>It is pipeline specific</li>
	<li>the message is for humans, not made for automation</li>
	<li>I could not find a way to only retrieve queued items. There seems to be a way for that given here <a href="https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-rest#find-latest-run-on-all-pipelines" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-rest#find-latest-run-on-all-pipelines</a> (and adapting with <tt>status:QUEUED</tt> or something), but {<tt>JENKINS_URL}/blue/rest/</tt> gives me a 404.</li>
	<li>If relies on Blue Ocean to be installed</li>
</ol>

<p> </p>
<h1><a name="WhatIpropose%28apossiblesolution%29"></a>What I propose (a possible solution)</h1>

<p>When accessing {<tt>JENKINS_URL}/queue/api/json/ :</tt></p>

<p><tt>A new key:</tt></p>

<pre>
<span class="code-quote">"requires"</span>: [
    {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.operators.Operator$OR"</span>,
        <span class="code-quote">"operands"</span>: [
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"agent"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"for_xxx"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"other_label"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.operators.Operator$AND"</span>,
                <span class="code-quote">"operands"</span>: [
                {
                    <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Node"</span>,
                    <span class="code-quote">"name"</span>: <span class="code-quote">"build2"</span>
                },
                {
                    <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                    <span class="code-quote">"name"</span>: <span class="code-quote">"debug"</span>
                },
                ]
            }
        ]
    }
]
</pre>
<p>That would explain what is required in order to unblock the queued item. In the above example, that would mean "an executor on a computer with a label 'agent', 'for_xxx' or 'other_label' or on a computer named 'build2' with label 'debug'".</p>

<p>The grammar is obviously just a draft/proposal and would need some reflection first.</p>

<p>But the goal would be (independently on the solution taken) to know what the job needs in order to run to make it easier (or at least reliable) to provision resource for jenkins.</p>

<p> </p>

<p> </p>

<p> </p>

<p> </p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mulugruntz.png" title="mulugruntz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mulugruntz">mulugruntz</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51198" target="_blank">Jenkins' CauseOfBlockage through the API</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hi,</p>

<p> </p>

<p>It may sound like a duplicate of <del><a href="https://issues.jenkins.io/browse/JENKINS-41579" title="API for Pipeline level for causeOfBlockage" class="issue-link" data-issue-key="JENKINS-41579"><del>JENKINS-41579</del></a></del> but it's not.</p>

<p>It's currently not possible to have the <tt>CauseOfBlockage</tt> through the json API.</p>

<p>There are currently three half-baked ways to do this, but none are usable for me.</p>

<p>To reproduce, create a pipeline with the following groovy script:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node(<span class="code-quote">'build3'</span>) { <span class="code-comment">// change <span class="code-keyword">for</span> a name that you possess and one that you don't
</span>  stage(<span class="code-quote">'test'</span>) {
    sleep 5
  }
}
</pre>
</div></div>
<h1><a name="Firstway"></a>First way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/queue/api/json</tt> , it shows (among other) the following piece of data:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">{
      <span class="code-quote">"_class"</span>: <span class="code-quote">"hudson.model.Queue$BuildableItem"</span>,
      <span class="code-quote">"actions"</span>: [
        {}
      ],
      <span class="code-quote">"blocked"</span>: <span class="code-keyword">false</span>,
      <span class="code-quote">"buildable"</span>: <span class="code-keyword">true</span>,
      <span class="code-quote">"id"</span>: 264,
      <span class="code-quote">"inQueueSince"</span>: 1525768721244,
      <span class="code-quote">"params"</span>: "",
      <span class="code-quote">"stuck"</span>: <span class="code-keyword">true</span>,
      <span class="code-quote">"task"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution$PlaceholderTask"</span>
      },
      <span class="code-quote">"url"</span>: <span class="code-quote">"queue/item/264/"</span>,
      <span class="code-quote">"why"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
      <span class="code-quote">"buildableStartMilliseconds"</span>: 1525768721245,
      <span class="code-quote">"pending"</span>: <span class="code-keyword">false</span>
},
</pre>
</div></div>
<p>The other <tt>why</tt> messages could be, for example:
 "why": "build2 is offline",
  </p>

<p>I believe all possible messages are located there: <a href="https://github.com/jenkinsci/jenkins/tree/1feb23e55ca807a664cf89986241155f24038aa5/core/src/main/resources/hudson/model/queue/CauseOfBlockage" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/tree/1feb23e55ca807a664cf89986241155f24038aa5/core/src/main/resources/hudson/model/queue/CauseOfBlockage</a></p>
<h3><a name="Theproblem"></a>The problem</h3>

<p>I cannot use this information for 2 reasons:</p>
<ol>
	<li>it is language dependent</li>
	<li><tt>BECAUSE_LABEL_IS_BUSY</tt> and <tt>BECAUSE_NODE_IS_BUSY</tt> result in the same string</li>
	<li>the message is for humans, not made for automation</li>
</ol>

<h1><a name="Secondway"></a>Second way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/{pipeline_name}/queue/</tt> , it should show the piece of data given in <a href="https://issues.jenkins-ci.org/browse/JENKINS-41579" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-41579</a> .</p>

<p>When I do it with my current queue (4 items in the queue), I see this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[]
</pre>
</div></div>
<p>Not really useful, eh?</p>

<p>And with another pipeline, I see (among other lines) this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"Build #160 is already in progress (ETA:N/A)"</span>,
</pre>
</div></div>
<h3><a name="Theproblem"></a>The problem</h3>

<p>I think this one is self explanatory. However, there are other issues to it:</p>
<ol>
	<li>All queued items don't appear</li>
	<li>When it does appear, the message is for humans, not made for automation.</li>
	<li>It is pipeline specific. I couldn't find a way to have this info for all pipelines (or only the latest run within each pipeline)</li>
	<li>It relies on Blue Ocean to be installed</li>
</ol>

<p> </p>
<h1><a name="Thirdway"></a>Third way</h1>
<h3><a name="Howtodoit"></a>How to do it</h3>

<p>When accessing  {<tt>JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/</tt>{<tt>pipeline_name</tt>}<tt>/runs/</tt> , it shows (among other) the following piece of data:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">  {
    <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.impl.pipeline.PipelineRunImpl"</span>,
    <span class="code-quote">"_links"</span>: {
      <span class="code-quote">"parent"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/"</span>
      },
      <span class="code-quote">"tests"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/tests/"</span>
      },
      <span class="code-quote">"nodes"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/nodes/"</span>
      },
      <span class="code-quote">"log"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/log/"</span>
      },
      <span class="code-quote">"self"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/"</span>
      },
      <span class="code-quote">"actions"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/actions/"</span>
      },
      <span class="code-quote">"steps"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/steps/"</span>
      },
      <span class="code-quote">"artifacts"</span>: {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.<span class="code-keyword">rest</span>.hal.Link"</span>,
        <span class="code-quote">"href"</span>: <span class="code-quote">"/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/<span class="code-object">short</span> pipeline/runs/87/artifacts/"</span>
      }
    },
    <span class="code-quote">"actions"</span>: [],
    <span class="code-quote">"artifactsZipFile"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
    <span class="code-quote">"causes"</span>: [
      {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"io.jenkins.blueocean.service.embedded.<span class="code-keyword">rest</span>.AbstractRunImpl$BlueCauseImpl"</span>,
        <span class="code-quote">"shortDescription"</span>: <span class="code-quote">"Started by user XXX"</span>,
        <span class="code-quote">"userId"</span>: <span class="code-quote">"xxx"</span>,
        <span class="code-quote">"userName"</span>: <span class="code-quote">"XXX"</span>
      }
    ],
    <span class="code-quote">"changeSet"</span>: [],
    <span class="code-quote">"description"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"durationInMillis"</span>: 0,
    <span class="code-quote">"enQueueTime"</span>: <span class="code-quote">"2018-05-08T08:38:37.482+0000"</span>,
    <span class="code-quote">"endTime"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"estimatedDurationInMillis"</span>: 102598,
    <span class="code-quote">"id"</span>: <span class="code-quote">"87"</span>,
    <span class="code-quote">"name"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"organization"</span>: <span class="code-quote">"jenkins"</span>,
    <span class="code-quote">"pipeline"</span>: <span class="code-quote">"{pipeline_name}"</span>,
    <span class="code-quote">"replayable"</span>: <span class="code-keyword">true</span>,
    <span class="code-quote">"result"</span>: <span class="code-quote">"UNKNOWN"</span>,
    <span class="code-quote">"runSummary"</span>: <span class="code-quote">"?"</span>,
    <span class="code-quote">"startTime"</span>: <span class="code-quote">"2018-05-08T08:38:37.486+0000"</span>,
    <span class="code-quote">"state"</span>: <span class="code-quote">"QUEUED"</span>,
    <span class="code-quote">"type"</span>: <span class="code-quote">"WorkflowRun"</span>,
    <span class="code-quote">"branch"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"commitId"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"commitUrl"</span>: <span class="code-keyword">null</span>,
    <span class="code-quote">"pullRequest"</span>: <span class="code-keyword">null</span>
  },
</pre>
</div></div>
<p>We can see the following line:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"causeOfBlockage"</span>: <span class="code-quote">"There are no nodes with the label ‘build3’"</span>,
</pre>
</div></div>
<p>So, here we have the information again.</p>
<h3><a name="Theproblem"></a>The problem</h3>

<p>It comes with several problems:</p>
<ol>
	<li>It is pipeline specific</li>
	<li>the message is for humans, not made for automation</li>
	<li>I could not find a way to only retrieve queued items. There seems to be a way for that given here <a href="https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-rest#find-latest-run-on-all-pipelines" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-rest#find-latest-run-on-all-pipelines</a> (and adapting with <tt>status:QUEUED</tt> or something), but {<tt>JENKINS_URL}/blue/rest/</tt> gives me a 404.</li>
	<li>If relies on Blue Ocean to be installed</li>
</ol>

<p> </p>
<h1><a name="WhatIpropose%28apossiblesolution%29"></a>What I propose (a possible solution)</h1>

<p>When accessing {<tt>JENKINS_URL}/queue/api/json/ :</tt></p>

<p><tt>A new key:</tt></p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"requires"</span>: [
    {
        <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.operators.Operator$OR"</span>,
        <span class="code-quote">"operands"</span>: [
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"agent"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"for_xxx"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                <span class="code-quote">"name"</span>: <span class="code-quote">"other_label"</span>
            },
            {
                <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.operators.Operator$AND"</span>,
                <span class="code-quote">"operands"</span>: [
                {
                    <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Node"</span>,
                    <span class="code-quote">"name"</span>: <span class="code-quote">"build2"</span>
                },
                {
                    <span class="code-quote">"_class"</span>: <span class="code-quote">"org.jenkinsci.[...].Label"</span>,
                    <span class="code-quote">"name"</span>: <span class="code-quote">"debug"</span>
                },
                ]
            }
        ]
    }
]
</pre>
</div></div>
<p>That would explain what is required in order to unblock the queued item. In the above example, that would mean "an executor on a computer with a label 'agent', 'for_xxx' or 'other_label' or on a computer named 'build2' with label 'debug'".</p>

<p>The grammar is obviously just a draft/proposal and would need some reflection first.</p>

<p>But the goal would be (independently on the solution taken) to know what the job needs in order to run to make it easier (or at least reliable) to provision resource for jenkins.</p>

<p> </p>

<p> </p>

<p> </p>

<p> </p></pre>
</details>
</details>
